File: apps/ios/Orange Cloud/Orange Cloud/Models/R2CatalogModels.swift
//
//  R2CatalogModels.swift
//  Orange Cloud
//
//  R2 Data Catalog —— 把 R2 桶启用为 Apache Iceberg 目录。
//  /accounts/{account_id}/r2-catalog[/{bucket_name}]
//
//  2026-08-03 起计费（$9/百万次目录操作，每月含 100 万次免费；压实另计），
//  UI 上要让用户知道启用是有成本的动作。
//

import Foundation

nonisolated struct R2Catalog: Codable, Sendable {
    let id:                String?
    /// 所属桶名
    let bucket:            String?
    /// 完整目录名，形如 account123_analytics-bucket
    let name:              String?
    /// active / …
    let status:            String?
    /// present 表示已存过访问凭据
    let credentialStatus:  String?
    let maintenanceConfig: R2CatalogMaintenance?

    enum CodingKeys: String, CodingKey {
        case id, bucket, name, status
        case credentialStatus  = "credential_status"
        case maintenanceConfig = "maintenance_config"
    }

    var isActive: Bool { status == "active" }
}

nonisolated struct R2CatalogMaintenance: Codable, Sendable {
    let compaction:         R2CatalogCompaction?
    let snapshotExpiration: R2CatalogSnapshotExpiration?

    enum CodingKeys: String, CodingKey {
        case compaction
        case snapshotExpiration = "snapshot_expiration"
    }
}

nonisolated struct R2CatalogCompaction: Codable, Sendable {
    let state:        String?
    /// 接口返回的是字符串数字（如 "128"）
    let targetSizeMb: String?

    enum CodingKeys: String, CodingKey {
        case state
        case targetSizeMb = "target_size_mb"
    }
}

nonisolated struct R2CatalogSnapshotExpiration: Codable, Sendable {
    let state:              String?
    let maxSnapshotAge:     String?
    let minSnapshotsToKeep: Int?

    enum CodingKeys: String, CodingKey {
        case state
        case maxSnapshotAge     = "max_snapshot_age"
        case minSnapshotsToKeep = "min_snapshots_to_keep"
    }
}

/// 命名空间列表的 result 外壳
nonisolated struct R2CatalogNamespaceList: Codable, Sendable {
    let namespaces: [R2CatalogNamespace]?
}

nonisolated struct R2CatalogNamespace: Codable, Identifiable, Hashable, Sendable {
    /// Iceberg 命名空间是分段的，接口以数组返回
    let name: [String]?

    var id: String { displayName }

    /// 嵌套命名空间用 · 连接展示
    var displayName: String {
        (name ?? []).joined(separator: " · ")
    }

    /// SQL 里引用表用的前缀（嵌套命名空间以 . 连接）
    var sqlName: String {
        (name ?? []).joined(separator: ".")
    }
}

/// 表清单的 result 外壳（GET .../namespaces/{ns}/tables）
nonisolated struct R2CatalogTableList: Codable, Sendable {
    let identifiers: [R2CatalogTableIdentifier]?
}

nonisolated struct R2CatalogTableIdentifier: Codable, Identifiable, Hashable, Sendable {
    let name:      String?
    let namespace: [String]?

    var id: String { sqlName }

    /// SQL 里的完整引用：namespace.table
    var sqlName: String {
        ((namespace ?? []) + [name ?? ""]).filter { !$0.isEmpty }.joined(separator: ".")
    }
}

File: apps/android/app/src/main/kotlin/jiamin/chen/orangecloud/data/model/R2CatalogModels.kt
package jiamin.chen.orangecloud.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * R2 Data Catalog —— 把 R2 桶启用为 Apache Iceberg 目录。
 * /accounts/{account_id}/r2-catalog[/{bucket_name}]
 *
 * 2026-08-03 起计费（$9/百万次目录操作，每月含 100 万次免费），
 * UI 上要让用户知道启用是有成本的动作。
 */
@Serializable
data class R2Catalog(
    val id: String? = null,
    val bucket: String? = null,
    /** 完整目录名，形如 account123_analytics-bucket */
    val name: String? = null,
    val status: String? = null,
    /** present 表示已存过访问凭据 */
    @SerialName("credential_status") val credentialStatus: String? = null,
    @SerialName("maintenance_config") val maintenanceConfig: R2CatalogMaintenance? = null,
) {
    val isActive: Boolean get() = status == "active"
}

@Serializable
data class R2CatalogMaintenance(
    val compaction: R2CatalogCompaction? = null,
    @SerialName("snapshot_expiration") val snapshotExpiration: R2CatalogSnapshotExpiration? = null,
)

@Serializable
data class R2CatalogCompaction(
    val state: String? = null,
    /** 接口返回的是字符串数字（如 "128"） */
    @SerialName("target_size_mb") val targetSizeMb: String? = null,
)

@Serializable
data class R2CatalogSnapshotExpiration(
    val state: String? = null,
    @SerialName("max_snapshot_age") val maxSnapshotAge: String? = null,
    @SerialName("min_snapshots_to_keep") val minSnapshotsToKeep: Int? = null,
)

/** 命名空间列表的 result 外壳 */
@Serializable
data class R2CatalogNamespaceList(val namespaces: List<R2CatalogNamespace>? = null)

@Serializable
data class R2CatalogNamespace(
    /** Iceberg 命名空间是分段的，接口以数组返回 */
    val name: List<String>? = null,
) {
    /** 嵌套命名空间用 · 连接展示 */
    val displayName: String get() = name.orEmpty().joinToString(" · ")

    /** SQL 里引用表用的前缀（嵌套命名空间以 . 连接） */
    val sqlName: String get() = name.orEmpty().joinToString(".")
}

/** 表清单的 result 外壳（GET .../namespaces/{ns}/tables） */
@Serializable
data class R2CatalogTableList(val identifiers: List<R2CatalogTableIdentifier>? = null)

@Serializable
data class R2CatalogTableIdentifier(
    val name: String? = null,
    val namespace: List<String>? = null,
) {
    /** SQL 里的完整引用：namespace.table */
    val sqlName: String
        get() = (namespace.orEmpty() + listOfNotNull(name)).filter { it.isNotEmpty() }.joinToString(".")
}
